nixos/modules/services/monitoring/prometheus/default.nix

```diff
@@ -1830,40 +1778,26 @@ in
       uid = config.ids.uids.prometheus;
       group = "prometheus";
     };
+    environment.etc."prometheus/prometheus.yaml" = mkIf cfg.enableReload {
+      source = prometheusYml;
+    };
     systemd.services.prometheus = {
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
-      preStart = mkIf (!cfg.enableReload) ''
-        ${lib.getBin pkgs.envsubst}/bin/envsubst -o "/run/prometheus/prometheus-substituted.yaml" \
-                                                 -i "${prometheusYml}"
-      '';
       serviceConfig = {
         ExecStart = "${cfg.package}/bin/prometheus" +
           optionalString (length cmdlineArgs != 0) (" \\\n  " +
             concatStringsSep " \\\n  " cmdlineArgs);
         ExecReload = mkIf cfg.enableReload "+${reload}/bin/reload-prometheus";
         User = "prometheus";
         Restart = "always";
-        EnvironmentFile = mkIf (cfg.environmentFile != null && !cfg.enableReload) [ cfg.environmentFile ];
```

Comment:
Should there maybe a deprecation phase?

Reply:
Since this is causing issues (https://github.com/NixOS/nixpkgs/issues/126083, https://github.com/NixOS/nixpkgs/pull/127889#issuecomment-869185960) and since there's now an alternative I think it's better to go without a deprecation cycle. I did now add a `mkRemovedOptionModule` so that users get a helpful error message when they define `services.prometheus.environmentFile`.